Привет, меня зовут Артем Хуршудов. Сегодня я проверю твой проект.
<br> Дальнейшее общение будет происходить на "ты" если это не вызывает никаких проблем.
<br> Желательно реагировать на каждый мой комментарий ('исправил', 'не понятно как исправить ошибку', ...)
<br> Пожалуйста, не удаляй комментарии ревьюера, так как они повышают качество повторного ревью.

Комментарии будут в <font color='green'>зеленой</font>, <font color='blue'>синей</font> или <font color='red'>красной</font> рамках:

<div class="alert alert-block alert-success">
<b>Успех:</b> Если все сделано отлично
</div>

<div class="alert alert-block alert-info">
<b>Совет: </b> Если можно немного улучшить
</div>

<div class="alert alert-block alert-danger">
<b>Ошибка:</b> Если требуются исправления. Работа не может быть принята с красными комментариями.
</div>

-------------------

Будет очень хорошо, если ты будешь помечать свои действия следующим образом:
<div class="alert alert-block alert-warning">
<b>Комментарий студента:</b> ...
</div>

<div class="alert alert-block alert-warning">
<b>Изменения:</b> Были внесены следующие изменения ...
</div>

### <font color='orange'>Общее впечатление</font>
* Большое спасибо за проделанную работу. Видно, что приложено много усилий.
* Отлично, что правильно использованы train/valid/test наборы.
* Выводы и рассуждения получились содержательными, их было интересно читать.
* Я оставил несколько советов, надеюсь, что они будут учтены в будущих проектах.
* Тебе удалось добиться очень хорошего качества, поздравляю!
* Проект может быть зачтен, но я его отправлю назад, чтобы у тебя была возможность задать вопросы и внести правки, при желании. Однако, ты можешь просто вернуть проект в таком же виде и я его зачту.


### <font color='orange'>Общее впечатление (ревью 2)</font>
* Проект зачтен!
* Для удобства все новые комментарии обозначены фразой "ревью 2".
* Удачи в дальнейшем обучении и следующих работах!

# Рекомендация тарифов

Оператор мобильной связи «Мегалайн» выяснил: многие клиенты пользуются архивными тарифами. Они хотят построить систему, способную проанализировать поведение клиентов и предложить пользователям новый тариф: «Смарт» или «Ультра».
В качестве источника информации были выбраны данные о поведении клиентов, которые уже перешли на эти тарифы.

**Цель исследования** - построить модель для задачи классификации, которая выберет подходящий тариф пользователям.

**Ход исследования**

Построим модель с максимально большим значением accuracy. Долю правильных ответов доведем по крайней мере до 0.75. Данные о поведении клиентов взяты из файла users_behavior.csv. Данные уже прошли предобработку.

Таким образом, исследование можно разделить на следующие этапы:

1. Изучение данных из файла.
2. Разбивка данных на выборки.
3. Исследование моделей.
4. Проверка модели на тестовой выборке.
5. Проверка модели на адекватность.

# Содержание
* [Изучение данных из файла](#Изучение-данных-из-файла)
* [Разбивка данных на выборки](#Разбивка-данных-на-выборки)
* [Исследование моделей](#Исследование-моделей)
* [Проверка модели на тестовой выборке](#Проверка-модели-на-тестовой-выборке)
* [Проверка модели на адекватность](#Проверка-модели-на-адекватность)

## Откройте и изучите файл

In [14]:
# импортируем все необходимые для исследования библиотеки
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.dummy import DummyClassifier

<div class="alert alert-block alert-success">
<b>Успех:</b> Отлично, что все импорты собраны в первой ячейке ноутбука! Если у того, кто будет запускать твой ноутбук будут отсутствовать некоторые библиотеки, то он это увидит сразу, а не в процессе!
</div>

In [15]:
# загрузим файл с данными
df = pd.read_csv('/datasets/users_behavior.csv')

In [16]:
# выведем на экран первые 5 строк df
df.head()

calls  minutes  messages   mb_used  is_ultra
0   40.0   311.90      83.0  19915.42         0
1   85.0   516.75      56.0  22696.96         0
2   77.0   467.66      86.0  21060.45         0
3  106.0   745.53      81.0   8437.39         1
4   66.0   418.74       1.0  14502.75         0

In [17]:
# получим общую информацию о данных в таблице df
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3214 entries, 0 to 3213
Data columns (total 5 columns):
calls       3214 non-null float64
minutes     3214 non-null float64
messages    3214 non-null float64
mb_used     3214 non-null float64
is_ultra    3214 non-null int64
dtypes: float64(4), int64(1)
memory usage: 125.7 KB


Итак, таблица состоит из 3214 объектов и 5 признаков.

Согласно документации к данным:

- сalls — количество звонков;
- minutes — суммарная длительность звонков в минутах;
- messages — количество sms-сообщений;
- mb_used — израсходованный интернет-трафик в Мб;
- is_ultra — каким тарифом пользовался в течение месяца («Ультра» — 1, «Смарт» — 0).
Целевым признаком является is_ultra, 

<div class="alert alert-block alert-success">
<b>Успех:</b> Данные загружены корректно, первичный осмотр проведен.
</div>

## Разбивка данных на выборки

Чтобы разбить данные на 3 выборки, сперва произведем разбивку на 2 выборки - обучающую и тестовую (60% и 40% соответственно) с помощью train_test_split. Затем аналогичным способом выделим половину из тестовой выборки для создания валидационной.

In [18]:
# разбивка на обучающую и тестовую выборки
df_train, df_test = train_test_split(df, test_size=0.4)

In [19]:
# разбивка на тестовую и валидационную выборки
df_test, df_valid = train_test_split(df_test, test_size=0.5)

In [20]:
# выделим в каждой выборке признаки и целевой признак для обучения модели
# обучающая выборка
features_train = df_train.drop(['is_ultra'], axis=1)
target_train = df_train['is_ultra']
# валидационная выборка
features_valid = df_valid.drop(['is_ultra'], axis=1)
target_valid = df_valid['is_ultra']
# тестовая выборка
features_test = df_test.drop(['is_ultra'], axis=1)
target_test = df_test['is_ultra']

In [21]:
# определим размеры полученных наборов для избежания ошибок
print('Количество строк в обучающей выборке features_train:', len(features_train))
print('Количество строк в обучающей выборке target_train:', len(target_train))
print('Количество строк в валидационной выборке features_valid:', len(features_valid))
print('Количество строк в валидационной выборке target_valid:', len(target_valid))
print('Количество строк в тестовой выборке features_test:', len(features_test))
print('Количество строк в тестовой выборке target_test:', len(target_test))

Количество строк в обучающей выборке features_train: 1928
Количество строк в обучающей выборке target_train: 1928
Количество строк в валидационной выборке features_valid: 643
Количество строк в валидационной выборке target_valid: 643
Количество строк в тестовой выборке features_test: 643
Количество строк в тестовой выборке target_test: 643


В рамках данного этапа исследования разбили выборки на 3 группы - обучающую (60%), валидационную (20%) и тестовую (20%). Также были выделены признаки и целевой признак для дальнейшего обучения модели.

<div class="alert alert-block alert-success">
<b>Успех:</b> Разбиение было сделано верно.
</div>

<div class="alert alert-block alert-info">
<b>Совет: </b> Желательно было напечатать размеры полученных наборов. Иногда, это помогает избежать ошибок.
</div>

<div class="alert alert-block alert-warning">
<b>Комментарий студента:</b> Напечатал размеры полученных наборов.
</div>

<div class="alert alert-block alert-success">
<b>Успех (ревью 2):</b> Спасибо!
</div>

## Исследование моделей

На данном этапе сравним резульаты использования 3 моделей: решающего дерева, случайного леса и регрессии.

Начнем с решающего дерева. Найдем наилучший результат в зависимости от изменения гиперпараметра max_depth.

In [22]:
# найдем качество модели решающего дерева
best_tree_model = None
best_tree_result = 0
best_depth = 0
for depth in range(1, 50):
    tree_model = DecisionTreeClassifier(random_state=12345, max_depth=depth) # обучим модель с заданной глубиной дерева
    tree_model.fit(features_train, target_train) # обучение модели
    valid_predictions=tree_model.predict(features_valid) # получим предсказания модели
    tree_result = accuracy_score(target_valid, valid_predictions) # посчитаем качество модели
    if tree_result > best_tree_result:
        best_tree_model=tree_model
        best_tree_result=tree_result
        best_depth=depth
print('Accuracy лучшей модели:', best_tree_result)
print('Лучшая глубина дерева:', best_depth)

Accuracy лучшей модели: 0.80248833592535
Лучшая глубина дерева: 6


<div class="alert alert-block alert-success">
<b>Успех:</b> Подбор параметров был сделан правильно!
</div>

In [23]:
# перейдем к оценке случайного леса
best_forest_model = None
best_forest_result = 0
best_est = 0
best_depth = 0
for est in range(1, 50):
    for depth in range(1, 10):
        forest_model = RandomForestClassifier(random_state=12345, n_estimators=est, max_depth=depth) # обучим модель с заданным количеством деревьев
        forest_model.fit(features_train, target_train) # обучение модели
        valid_predictions = forest_model.predict(features_valid) # получим предсказания модели
        forest_result=accuracy_score(target_valid, valid_predictions) # посчитаем качество модели
        if forest_result > best_forest_result:
            best_forest_model=forest_model
            best_forest_result=forest_result
            best_est=est
            best_depth=depth
print('Accuracy лучшей модели:', best_forest_result)
print('Лучшее количество деревьев:', best_est)
print('Лучшая глубина дерева:', best_depth)

Accuracy лучшей модели: 0.8133748055987559
Лучшее количество деревьев: 32
Лучшая глубина дерева: 8


<div class="alert alert-block alert-info">
<b>Совет: </b> Тут можно было попробовать сразу несколько параметров.
</div>

<div class="alert alert-block alert-warning">
<b>Комментарий студента:</b> Добавил глубину дерева.
</div>

<div class="alert alert-block alert-success">
<b>Успех (ревью 2):</b> Отлично!
</div>

In [24]:
# перейдем к оценке логистической регрессии
regres_model = LogisticRegression(random_state=12345) # проведем инициализацию модели логистической регрессии
regres_model.fit(features_train, target_train) # обучим модель на тренировочной выборке
valid_predictions = regres_model.predict(features_valid) # получим предсказания модели
regres_result = accuracy_score(target_valid, valid_predictions) # посчитаем качество модели
print('Accuracy модели логистической регрессии на валидационной выборке:', regres_result)

Accuracy модели логистической регрессии на валидационной выборке: 0.749611197511664


/opt/conda/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:432: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)


Таким образом, модель решающего дерева и случайного леса показывают схожую точность. Логистическая модель выдает самый худший результат.

<div class="alert alert-block alert-success">
<b>Успех:</b> Этот шаг был сделан хорошо. Радует, что ты попробовал разные модели!
</div>

## Проверка модели на тестовой выборке

Оценим точность каждой модели на тестовой выборке.

In [25]:
# оценим точность каждой модели на тестовой выборке
test_predictions = best_tree_model.predict(features_test)
accuracy_tree = accuracy_score(target_test, test_predictions)
print('Точность модели решающего дерева на тестовой выборке',accuracy_tree)
test_predictions = best_forest_model.predict(features_test)
accuracy_forest = accuracy_score(target_test, test_predictions)
print('Точность модели случайного леса на тестовой выборке', accuracy_forest)
test_predictions = regres_model.predict(features_test)
accuracy_regres = accuracy_score(target_test, test_predictions)
print('Точность модели логистической регрессии на тестовой выборке', accuracy_regres)

Точность модели решающего дерева на тестовой выборке 0.7947122861586314
Точность модели случайного леса на тестовой выборке 0.8211508553654744
Точность модели логистической регрессии на тестовой выборке 0.7511664074650077


Точность моделей на тестовой выборке незначительно отличается от точности на валидационных данных, что говорит о непереобученности моделей. Самый высокий результат показывает модель случайного леса, но она имеет самую низкую скорость. Модель решающего дерева не сильно отстает в точности.

<div class="alert alert-block alert-success">
<b>Успех:</b> Тестирование было сделано абсолютно верно!
</div>

## Проверка модели на адекватность

In [26]:
# сравним точность модели случайного леса с Dummy моделью
dummy_clf = DummyClassifier(strategy='most_frequent', random_state=12345)
dummy_clf.fit(features_train, target_train)
dummy_score = dummy_clf.score(features_test, target_test)
print('Точность Dummy модели', dummy_score)
print('Точность модели случайного леса', accuracy_forest)

Точность Dummy модели 0.7091757387247278
Точность модели случайного леса 0.8211508553654744


Поскольку модель случайного леса лучше предсказывает, чем Dummy модель, то нашу модель можно считать адекватной.

<div class="alert alert-block alert-success">
<b>Успех:</b> Молодец, что здесь использовал готовую дамми модель!
</div>

## Вывод 

В процессе исследования мы составляли оптимальную модель для предсказания пользователям "Мегалайн" оптимального тарифа. Было рассмотрено 3 модели с различными гиперпараметрами: модель решающего дерева, модель случайного леса и модель логистической регрессии. Модель случайного леса показала наибольшую точность предсказаний. Также данная модель была признана адекватной.

<div class="alert alert-block alert-success">
<b>Успех:</b> Приятно видеть вывод в конце проекта!
</div>

## Чек-лист готовности проекта

Поставьте 'x' в выполненных пунктах. Далее нажмите Shift+Enter.

- [x] Jupyter Notebook открыт
- [x] Весь код исполняется без ошибок
- [x] Ячейки с кодом расположены в порядке исполнения
- [x] Выполнено задание 1: данные загружены и изучены
- [x] Выполнено задание 2: данные разбиты на три выборки
- [x] Выполнено задание 3: проведено исследование моделей
    - [x] Рассмотрено больше одной модели
    - [x] Рассмотрено хотя бы 3 значения гипепараметров для какой-нибудь модели
    - [x] Написаны выводы по результатам исследования
- [x] Выполнено задание 3: Проведено тестирование
- [x] Удалось достичь accuracy не меньше 0.75
